flatfs.go

```diff
@@ -404,6 +597,172 @@ func (fs *Datastore) walkTopLevel(path string, reschan chan query.Result) error
 	return nil
 }
 
+// folderSize estimates the diskUsage of a folder by reading
+// up to DiskUsageFilesAverage entries in it and assumming any
+// other files will have an avereage size.
+func folderSize(path string) (int64, error) {
+	var du int64
+
+	folder, err := os.Open(path)
+	if err != nil {
+		return 0, err
+	}
+	defer folder.Close()
+
+	stat, err := folder.Stat()
+	if err != nil {
+		return 0, err
+	}
+
+	files, err := folder.Readdirnames(0)
+	if err != nil {
+		return 0, err
+	}
+
+	totalFiles := len(files)
+	filesToProcess := files
+	nonProcessed := 0
+	if DiskUsageFilesAverage < totalFiles && DiskUsageFilesAverage > 0 {
+		nonProcessed = totalFiles - DiskUsageFilesAverage
+		filesToProcess = files[0:DiskUsageFilesAverage]
```

Comment:
We should pick those randomly, as the order in which the files are returned may correspond to the order they were added in (this as quite system-specific, but may happen in theory), and thus won't get us the uniform size distribution.